### PR TITLE
Exp cleanup

### LINKED ISSUE #4893: Exponentiation producing inconsistent results
Across multiple unsigned integer types I've tried - e.g. uint8 - the exponentiation operator can produce a result which is outside of the range of that type in some contexts, but exhibits the truncation behavior I'd expect in other contexts.

I get different failure modes in the truffle-based test framework I'm using (with ganache-cli, Linux, solc-js 0.4.24 - this produces unexpected values) and http://remix.ethereum.org/ (this produces a VM error).

See:

    contract c0 {
      // OK: 0 ** 0 returns 1
      function f1() returns (uint8) { return uint8(0) ** uint8(0); }

      // FAILURE:
      //   - In truffle: returns 0, indicating as per above that
      // "uint8(uint8(2) ** uint8(8))" is not 0 - but it should be as per the uint8 range 
      //   - In Remix: transaction fails due to a VM eror
      function f2() returns (uint8) { return uint8(0) ** uint8(uint8(2) ** uint8(8)); }

      // OK: 2 ** 8 returns the expected zero-truncated result
      function f3() returns (uint8) { return uint8(2) ** uint8(8); }

      // OK: In this context, "uint8(uint8(2) ** uint8(8))" is 0, contradicting
      // the exponentiation context above
      function f4() returns (bool) { return uint8((uint8(2) ** uint8(8))) > 0; }
    }

This probably has something to do with this statement from the docs, which indicates that intermediate results exceeding the size of their input operand types can be produced:

	https://solidity.readthedocs.io/en/v0.4.24/types.html

	Number literal expressions retain arbitrary precision until they are converted
	to a non-literal type (i.e. by using them together with a non-literal expression).
	This means that computations do not overflow and divisions do not truncate in
	number literal expressions.

	For example, (2**800 + 1) - 2**800 results in the constant 1 (of type uint8) 
	although intermediate results would not even fit the machine word size.

If I understand this correctly, the expression "x ** y" is dynamically typed and may  produce a result which is outside of the range of x's type. But in that case using an explicit conversion to the smaller type should at least consistently always truncate (preferred, IMO) or never truncate.

(This report is part of an ICE Center@ETH Zurich project on automated compiler validation funded by the Ethereum Foundation.)

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #4893

### CODE REVIEW

**Comment by chriseth** [comment]
This might warrant an intermediate bugfix release.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -2129,7 +2129,9 @@ bool ExpressionCompiler::cleanupNeededForOp(Type::Category _type, Token::Value _
 {
 	if (Token::isCompareOp(_op) || Token::isShiftOp(_op))
 		return true;
-	else if (_type == Type::Category::Integer && (_op == Token::Div || _op == Token::Mod))
+	else if (_type == Type::Category::Integer && (_op == Token::Div || _op == Token::Mod || _op == Token::Exp))
+		// We need cleanup for EXP because 0**0 == 1, but 0**0x100 == 0
+		// It would suffice to clean the oponent, though.
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Spellchecker found `oponent  ==> opponent`.
Is it a typo though? (exponent)

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Good old spelly!

**Review by ekpyron — DISMISSED** [review]

File: docs/bugs.json [hunk 2]
```diff
@@ -1,4 +1,11 @@
 [
+	{
+		"name": "ExpCleanupZero",
+		"summary": "Using exponentiating with types shorter than 256 bits and zero as base can result in unexpected values.",
+		"description": "As defined by the EVM, 0 ** 0 is one. Solidity does not always remove dirty higher order bits of types that are smaller than 256 bits before applying an operation because it does not matter if the cleanup is performed before or after the operation. This is true for EXP except for the case where the base is zero, which was overlooked.",
+		"fixed": "0.5.0",
```
Comment (chriseth) [inline docs/bugs.json]: Needs to be changed.

**Comment by chriseth** [comment]
We should also add a regex, which could maybe just be `**`, but should exclude `/**`

**Comment by axic** [comment]
> but should exclude `/**`

and `**/`

File: docs/bugs.json (hunk 2 above)
Comment (chriseth) [reply docs/bugs.json]: Will do that with the cherry-picking to develop_v0425

Comment (chriseth) [reply docs/bugs.json]: Actually no, that doesn't make sense.

**Comment by chriseth** [comment]
Ah nice, the buglist regex test found a fatal flaw in my original regex :)

**Comment by chriseth** [comment]
Should be ready for final review.

**Comment by chriseth** [comment]
Rebased.

**Review by ekpyron — COMMENTED** [review]
Maybe the test cases could also verify cleanup for exponentiation with bases other than ``0``. Otherwise looks fine.

**Comment by chriseth** [comment]
Added another test and reduced severity to "medium/high" after the observation that cleanup is properly performed if the type of the base is wider than the type of the exponent.

File: docs/bugs.json [hunk 3]
```diff
@@ -1,4 +1,12 @@
 [
+    {
+        "name": "ExpExponentCleanup",
+        "summary": "Using the ** operator with an exponent of type shorter than 256 bits can result in unexpected values.",
+        "description": "Higher order bits in the exponent are not properly cleaned before the EXP opcode is applied if the type of the exponent expression is smaller than 256 bits. In that case, the result might be larger than expected if the exponent is assumed to lie within the value range of the type. Literal numbers as exponents are unaffected as are variables of type uint256.",
```
Comment (ekpyron) [inline docs/bugs.json]: Should the description mentation that this does not happen, if the base type is wider than the exponent type?

Comment (chriseth) [reply docs/bugs.json]: Yes, I will add it.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 4]
```diff
@@ -11870,6 +11870,46 @@ BOOST_AUTO_TEST_CASE(shift_bytes_cleanup)
...
+	compileAndRun(sourceCode, 0, "C");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(u256(0x1)));
+}
+
+BOOST_AUTO_TEST_CASE(exp_cleanup_nonzero_base)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f() public pure returns (uint8 x) {
+				return uint8(0x166) ** uint8(uint8(2) ** uint8(8));
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp:11905]: What about cases like ``uint8(0x103) ** uint8(uint8(3) ** uint8(8))``?
Could either (correctly) be ``0x03 ^ 0xA1 == 0x03`` or (wrongly) ``0x3 ^ 0x19A1 == 0x53``...

Comment (ekpyron) [reply test/libsolidity/SolidityEndToEndTest.cpp:11905]: Maybe it's sufficient the way it is, though.

**Comment by chriseth** [comment]
Updated.

File: Changelog.md [hunk 5]
```diff
@@ -91,6 +91,9 @@ Compiler Features:
  * Code Generator: Allocate and free local variables according to their scope.
  * Removed ``pragma experimental "v0.5.0";``.
 
+Critical Bugfixes:
```
Comment (holiman) [inline Changelog.md:94]: If it's graded as `medium/high` might be better to use the wording "Important Bugfixes" instead of Critical

**Review by ekpyron — APPROVED** [review]

File: Changelog.md (hunk 5 above)
Comment (chriseth) [reply Changelog.md:94]: It will move to 0.4.25 anyway, so we have to change that later. I'll merge it for now.

**Comment by barakman** [comment]
@chriseth:

For those of us who haven't yet switched to version 0.4.25, where this problem is supposed to be resolved:

The bug description in your [release blog](https://blog.ethereum.org/2018/09/13/solidity-bugfix-release/), implies that if either the base or the exponent is/are 256 bits, then the operation is "safe".

Am I right to conclude this implication?

I have based it on the following statement (from the blog):

> the exponent of the `**` operator has a type that is shorter than 256 bits, but not shorter than the type of the base

Thanks

**Comment by chriseth** [comment]
If both types are 256 bit, then cleanup is a no-op, so it is safe to skip cleanup.

If one of the two operands has less than 256 bits but the other has 256 bits, then the combined type is a 256-bit integer. The operand that already has 256 bits cannot be "dirty" as explained above. The other operand has to be converted to a 256 bit integer. Cleanup is always done when an interger type is converted to a longer type, so the higher order bits of that operand are properly cleaned.

Conclusion: If either the base or the exponent has 256 bits (or both), the bug is not present.

**Pull request merged**